executor/memtable_reader.go

```diff
@@ -997,3 +997,116 @@ func (e *hotRegionsHistoryRetriver) getHotRegionRowWithSchemaInfo(
 	}
 	return row, nil
 }
+
+type tikvRegionPeersRetriever struct {
+	dummyCloser
+	extractor *plannercore.TikvRegionPeersExtractor
+	retrieved bool
+}
+
+func (e *tikvRegionPeersRetriever) retrieve(ctx context.Context, sctx sessionctx.Context) ([][]types.Datum, error) {
+	if e.extractor.SkipRequest || e.retrieved {
+		return nil, nil
+	}
+	e.retrieved = true
+	tikvStore, ok := sctx.GetStore().(helper.Storage)
+	if !ok {
+		return nil, errors.New("Information about hot region can be gotten only when the storage is TiKV")
+	}
+	tikvHelper := &helper.Helper{
+		Store:       tikvStore,
+		RegionCache: tikvStore.GetRegionCache(),
+	}
+
+	if len(e.extractor.StoreIDs) < 1 && len(e.extractor.RegionIDs) < 1 {
+		regionsInfo, err := tikvHelper.GetRegionsInfo()
+		if err != nil {
+			return nil, err
+		}
+		return e.packTiKVRegionPeersRows(regionsInfo.Regions)
+	}
+
+	var regionsInfo, regionsInfoByStoreID, regionsInfoByRegionID []helper.RegionInfo
+	regionMap := make(map[int64]struct{})
+	if len(e.extractor.StoreIDs) > 0 {
+		for _, storeID := range e.extractor.StoreIDs {
+			regionsInfo, err := tikvHelper.GetStoreRegionsInfo(storeID)
+			if err != nil {
+				return nil, err
+			}
+			// remove dup region
+			for _, region := range regionsInfo.Regions {
+				if _, ok := regionMap[region.ID]; !ok {
+					regionMap[region.ID] = struct{}{}
+					regionsInfoByStoreID = append(regionsInfoByStoreID, region)
+				}
+			}
+		}
+		if len(e.extractor.RegionIDs) < 1 {
+			return e.packTiKVRegionPeersRows(regionsInfoByStoreID)
+		}
```

Comment:
(1/3) I think this judgment is unnecessary since it will not enter the `len(e.extractor.RegionIDs) > 0` logic below either.

Reply:
Refactor and  move them before intersect.
The predicates extract by extractor is `and` case only, and   `len(regionsInfoByStoreID) == 0` may include two cases: 
1. no store_id predicate   **Do not intersect**
2. no corresponding store regionsInfo **Intersect**

so we should check the length of extracted predicate rather than result slice `regionsInfoByStoreID` or `regionsInfoByRegionID`.
```go
	// len(regionsInfoByStoreID) == 0 include two cases:
	// 1. no store_id predicate 2. no corresponding store regionsInfo
	if len(e.extractor.RegionIDs) == 0 {
		return e.packTiKVRegionPeersRows(regionsInfoByStoreID)
	}

	if len(e.extractor.StoreIDs) == 0 {
		return e.packTiKVRegionPeersRows(regionsInfoByRegionID)
	}

	// intersect
	for _, region := range regionsInfoByRegionID {
		if _, ok := regionMap[region.ID]; ok {
			regionsInfo = append(regionsInfo, region)
		}
	}
```

Reply:
Oh, I missed that part. Nice catch 👍. But what about the `or` condition 🤔?

Reply:
> Oh, I missed that part. Nice catch 👍. But what about the `or` condition 🤔?

```go
		if len(e.extractor.RegionIDs) < 1 {
			return e.packTiKVRegionPeersRows(regionsInfoByStoreID)
		}
```
In origin version implemention,the use of `regionsInfoByStoreID` and `regionsInfoByRegionID ` make `or`  difficult to use.

Reply:
How about the newest verion of implmention. I think we need to exclude unwanted regions got from `store_id` by `region_id` when specify both `store_id` and `region_id`, so I keep the `regionsInfoByStoreID`.